docs/smart-contracts/meebits_erc721.sol

```diff
@@ -0,0 +1,679 @@
+            // Compute the punk ID
+            uint punkId = _createVia.sub(1);
+            // Make sure the sender owns the punk
+            require(Cryptopunks(punks).punkIndexToAddress(punkId) == msg.sender, "Not the owner of this punk.");
+        }
+        creatorNftMints[_createVia]++;
+        return _mint(msg.sender, _createVia);
+    }
+
+    /**
+     * Public sale minting.
+     */
+    function mint() external payable reentrancyGuard returns (uint) {
+        require(publicSale, "Sale not started.");
+        require(!marketPaused);
+        require(numSales < SALE_LIMIT, "Sale limit reached.");
+        uint salePrice = getPrice();
+        require(msg.value >= salePrice, "Insufficient funds to purchase.");
+        if (msg.value > salePrice) {
+            msg.sender.transfer(msg.value.sub(salePrice));
+        }
+        beneficiary.transfer(salePrice);
+        numSales++;
+        return _mint(msg.sender, 0);
+    }
+
+    function _mint(address _to, uint createdVia) internal returns (uint) {
+        require(_to != address(0), "Cannot mint to 0x0.");
+        require(numTokens < TOKEN_LIMIT, "Token limit reached.");
+        uint id = randomIndex();
+
+        numTokens = numTokens + 1;
+        _addNFToken(_to, id);
+
+        emit Mint(id, _to, createdVia);
+        emit Transfer(address(0), _to, id);
+        return id;
+    }
+
+    function _addNFToken(address _to, uint256 _tokenId) internal {
+        require(idToOwner[_tokenId] == address(0), "Cannot add, already owned.");
+        idToOwner[_tokenId] = _to;
+
+        ownerToIds[_to].push(_tokenId);
+        idToOwnerIndex[_tokenId] = ownerToIds[_to].length.sub(1);
+    }
+
+    function _removeNFToken(address _from, uint256 _tokenId) internal {
+        require(idToOwner[_tokenId] == _from, "Incorrect owner.");
+        delete idToOwner[_tokenId];
+
+        uint256 tokenToRemoveIndex = idToOwnerIndex[_tokenId];
+        uint256 lastTokenIndex = ownerToIds[_from].length.sub(1);
+
+        if (lastTokenIndex != tokenToRemoveIndex) {
+            uint256 lastToken = ownerToIds[_from][lastTokenIndex];
+            ownerToIds[_from][tokenToRemoveIndex] = lastToken;
+            idToOwnerIndex[lastToken] = tokenToRemoveIndex;
+        }
+
+        ownerToIds[_from].pop();
+    }
+
+    function _getOwnerNFTCount(address _owner) internal view returns (uint256) {
+        return ownerToIds[_owner].length;
+    }
+
+    function _safeTransferFrom(address _from,  address _to,  uint256 _tokenId,  bytes memory _data) private canTransfer(_tokenId) validNFToken(_tokenId) {
+        address tokenOwner = idToOwner[_tokenId];
+        require(tokenOwner == _from, "Incorrect owner.");
+        require(_to != address(0));
+
+        _transfer(_to, _tokenId);
+
+        if (isContract(_to)) {
+            bytes4 retval = ERC721TokenReceiver(_to).onERC721Received(msg.sender, _from, _tokenId, _data);
+            require(retval == MAGIC_ON_ERC721_RECEIVED);
+        }
+    }
```

Comment:
Related to ERC721.safeTransformFrom
Does some checks and passes to _transfer
Some special logic related to if transfer is made to the contract